packages/services/src/view-service/fees.ts

```diff
@@ -0,0 +1,40 @@
+import { AssetId } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1/asset_pb';
+import {
+  TransactionPlannerRequest,
+  TransactionPlannerRequest_Output,
+  TransactionPlannerRequest_Swap,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1/view_pb';
+
+export const extractAltFee = (request: TransactionPlannerRequest): AssetId | undefined => {
+  const fields = [
+    { name: 'outputs', value: request.outputs },
+    { name: 'swaps', value: request.swaps },
+    { name: 'swapClaims', value: request.swapClaims },
+    { name: 'delegations', value: request.delegations },
+    { name: 'undelegations', value: request.undelegations },
+    { name: 'undelegationClaims', value: request.undelegationClaims },
+    { name: 'ibcRelayActions', value: request.ibcRelayActions },
+    { name: 'ics20Withdrawals', value: request.ics20Withdrawals },
+    { name: 'positionOpens', value: request.positionOpens },
+    { name: 'positionCloses', value: request.positionCloses },
+    { name: 'positionWithdraws', value: request.positionWithdraws },
+    { name: 'dutchAuctionScheduleActions', value: request.dutchAuctionScheduleActions },
+  ];
+
+  const nonEmptyField = fields.find(field => field.value.length > 0);
+
+  if (!nonEmptyField) {
+    console.warn('No non-empty field found in the request');
+    return undefined;
+  }
+
+  type PossibleTypes = TransactionPlannerRequest_Output | TransactionPlannerRequest_Swap;
+
+  const action = nonEmptyField.value[0] as PossibleTypes;
```

Comment:
Unless I'm missing something, this looks type-unsafe. We're iterating over all the possible fields of a request, which means `nonEmptyField` could contain _any_ type of actions, not just outputs and swaps. And the other action types might not have an `assetId` property. So, although the code we have here technically works for now, it could mislead future developers who will think that `action` is definitely either an output or a swap. I'd prefer instead to use the `.find()` above to find those specific action types so that the types are valid.

Reply:
good points, I [restricted](https://github.com/penumbra-zone/web/pull/1268/commits/89a24c183b4ab424250c0c08a35abb64086f1773) the scope to only supporting send / swap / auction actions. This means user's can only use alternative fees for that specified functionality, and returning `undefined` will cause the transaction to fail / prompt insufficient note balance because user doesn't have enough UM. Does that sufficiently address the concern here?

Reply:
> Unless I'm missing something, this looks type-unsafe.

good point, I [restricted](https://github.com/penumbra-zone/web/pull/1268/commits/89a24c183b4ab424250c0c08a35abb64086f1773) the scope to only supporting send / swap / auction actions. This means user's can only use alternative fees for that specified functionality, and returning `undefined` will cause the transaction to fail / prompt insufficient note balance because user doesn't have enough UM. Does that sufficiently address the concern here?